### ISSUE TITLE
expanded sdat files with all the species' cries in them do not play the title screen music

### ISSUE THREAD

**Issue by BluRosie** [BluRosie/hg-engine#47]
normally i would think this has to do with a somehow damaged sdat file, but now that we are using sdattool and that matches byte-for-byte i'm not 100% sure what the problem is

**Comment by BluRosie** [comment]
appears to have something to do with the filetable size - 2978 entries makes it fail as observed

**Comment by BluRosie** [comment]
+ global trade station + trainer battles

**Comment by BluRosie** [comment]
each sbnk appears to be loaded at the same time.  `0x5A20` is the difference between new cries and no cries in the sdat.  there are 412 new cries in the sdat--412 new sbnk's, each's data being `0x38` in length, which corresponds to the `0x5A20`.

**Comment by BluRosie** [comment]
PlayCry -> SndArcPlayerStartSeqEx_GBSounds -> NNS_SndArcPlayerStartSeqEx -> StartSeq -> NNSi_SndArcLoadBank -> SND_AssignWaveArc (!!!)

**Comment by BluRosie** [comment]
<details>
<summary>may not be SND_AssignWaveArc</summary>
<br>

```c
relevant structures:


typedef struct SNDBinaryFileHeader
{
    /* 0x0 */ char    signature[4];
    /* 0x4 */ u16     byteOrder;
    /* 0x6 */ u16     version;
    /* 0x8 */ u32     fileSize;
    /* 0xC */ u16     headerSize;
    /* 0xE */ u16     dataBlocks;
} // size = 0x10
SNDBinaryFileHeader;


typedef struct SNDBinaryBlockHeader
{
    /* 0x0 */ u32     kind;
    /* 0x4 */ u32     size;
} // size = 0x8
SNDBinaryBlockHeader;


// me when i recursive define but it's a pointer so it's okay
typedef struct SNDWaveArcLink
{
    /* 0x0 */ struct SNDWaveArc *waveArc;
    /* 0x4 */ struct SNDWaveArcLink *next;
} // size = 0x8
SNDWaveArcLink;


typedef struct SNDBankData
{
    /* 0x00 */ struct SNDBinaryFileHeader fileHeader;
    /* 0x10 */ struct SNDBinaryBlockHeader blockHeader;

    /* 0x18 */ struct SNDWaveArcLink waveArcLink[SND_BANK_TO_WAVEARC_MAX];
    /* 0x38 */ u32     instCount;
    /* 0x3C */ u32     instOffset[0];
} // size = 0x40
SNDBankData;


typedef struct SNDWaveArc
{
    /* 0x00 */ struct SNDBinaryFileHeader fileHeader;
    /* 0x10 */ struct SNDBinaryBlockHeader blockHeader;

    /* 0x18 */ struct SNDWaveArcLink *topLink;
    /* 0x1C */ u32     reserved[7];
    /* 0x38 */ u32     waveCount;
    /* 0x3C */ u32     waveOffset[0];
} // size = 0x40
SNDWaveArc;







/*---------------------------------------------------------------------------*
  Name:         SND_AssignWaveArc

  Description:  Assign waveform archives to banks (ThreadSafe)

  Arguments:    bank    - pointer to SNDBankData structure
                index   - wave archive index
                waveArc - pointer to SNDWaveArc structure

  Returns:      None.
 *---------------------------------------------------------------------------*/

void SND_AssignWaveArc(SNDBankData *bank, int index, SNDWaveArc *waveArc)
{
    SNDWaveArcLink *next;
    SNDWaveArcLink *prev;

    SNDi_LockMutex();

    if (bank->waveArcLink[index].waveArc != NULL)
    {
        // Already connected

        // If it is the same as the waveform archive you are trying to connect, do nothing
        if (waveArc == bank->waveArcLink[index].waveArc)
        {
            SNDi_UnlockMutex();
            return;
        }

        // If a different waveform archive is connected, disconnect it once.
        if (&bank->waveArcLink[index] == bank->waveArcLink[index].waveArc->topLink)
        {
            bank->waveArcLink[index].waveArc->topLink = bank->waveArcLink[index].next;

            DC_StoreRange(bank->waveArcLink[index].waveArc, sizeof(SNDWaveArc));
        }
        else
        {
            prev = bank->waveArcLink[index].waveArc->topLink;
            while (prev != NULL)
            {
                if (&bank->waveArcLink[index] == prev->next)
                    break;
                prev = prev->next;
            }
            prev->next = bank->waveArcLink[index].next;

            DC_StoreRange(prev, sizeof(SNDWaveArcLink));
        }
    }

    next = waveArc->topLink;
    waveArc->topLink = &bank->waveArcLink[index];
    bank->waveArcLink[index].next = next;
    bank->waveArcLink[index].waveArc = waveArc;

    SNDi_UnlockMutex();

    DC_StoreRange(bank, sizeof(SNDBankData));
    DC_StoreRange(waveArc, sizeof(SNDWaveArc));
}
```
</details>

**Comment by BluRosie** [comment]
<details>
<summary>NNSi_SndArcLoadBank shenanigans</summary>
<br>

```
/*---------------------------------------------------------------------------*
  Name:         NNSi_SndArcLoadBank

  Description:  Load bank data
 *---------------------------------------------------------------------------*/

NNSSndArcLoadResult NNSi_SndArcLoadBank(
    int bankNo,
    u32 loadFlag,
    NNSSndHeapHandle heap,
    BOOL bSetAddr,
    struct SNDBankData** pData
)
{
    const NNSSndArcBankInfo* bankInfo;
    const NNSSndArcWaveArcInfo* waveArcInfo;
    SNDBankData* bank = NULL;
    SNDWaveArc* waveArc;
    NNSSndArcLoadResult result;
    int i;

    // Get bank information
    bankInfo = NNS_SndArcGetBankInfo( bankNo );
    if ( bankInfo == NULL ) return NNS_SND_ARC_LOAD_ERROR_INVALID_BANK_NO;
    
    // If necessary to load
    if ( loadFlag & NNS_SND_ARC_LOAD_BANK )
    {
        bank = LoadBank( bankInfo->fileId, heap, bSetAddr );
        if ( bank == NULL ) {
            return NNS_SND_ARC_LOAD_ERROR_FAILED_LOAD_BANK;
        }
    }
    else
    {
        bank = (SNDBankData*)NNS_SndArcGetFileAddress( bankInfo->fileId );
    }
    
    // Load waveform data
    for( i = 0; i < NNS_SND_ARC_BANK_TO_WAVEARC_NUM ; i++ )
    {
        if ( bankInfo->waveArcNo[i] == NNS_SND_ARC_INVALID_WAVEARC_NO ) continue;
            
        // Get waveform archive information
        waveArcInfo = NNS_SndArcGetWaveArcInfo( bankInfo->waveArcNo[i] );
        if ( waveArcInfo == NULL ) return NNS_SND_ARC_LOAD_ERROR_INVALID_WAVEARC_NO;
        
        // Loading waveform archives
        result = NNSi_SndArcLoadWaveArc( bankInfo->waveArcNo[i], loadFlag, heap, bSetAddr, &waveArc );
        if ( result != NNS_SND_ARC_LOAD_SUCESS ) return result;
                
        if ( waveArcInfo->flags & NNS_SND_ARC_WAVEARC_SINGLE_LOAD )
        {
            // Individual waveform loading
            if ( loadFlag & NNS_SND_ARC_LOAD_WAVE )
            {
                if ( ! LoadSingleWaves( waveArc, bank, i, waveArcInfo->fileId, heap ) ) {
                    return NNS_SND_ARC_LOAD_ERROR_FAILED_LOAD_WAVE;
                }
            }
        }
            
        // Associate waveforms with banks
        if ( bank != NULL && waveArc != NULL ) {
            SND_AssignWaveArc( bank, i, waveArc );
        }
    }
    
    if ( pData != NULL ) *pData = bank;
    
    return NNS_SND_ARC_LOAD_SUCESS;
}
```
</details>

**Comment by BluRosie** [comment]
this has been significantly improved--now certain pokemon themselves seem to be causing it.  the title screen music loads in every time though

**Comment by BluRosie** [comment]
fixed by b4c827794f98a79b646e97a1c261167f3b4663ca for the most part

**Comment by BluRosie** [comment]
really truly fixed by f4e053e03a0840993b2d67ee8c74f4270519adcf
